### PR TITLE
executor: generate AST instead of SQL for foreign key cascade

### LINKED ISSUE #38795: use generate ast instead of SQL for foreign key cascade
## Enhancement

as the comment said: https://github.com/pingcap/tidb/pull/38652#pullrequestreview-1162117795

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs520 <crazycs520@gmail.com>

### What problem does this PR solve?

Issue Number: close #38795

Problem Summary: This PR use generate AST instead of SQL for foreign key cascade

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/foreign_key.go [hunk 1]
```diff
@@ -643,51 +647,37 @@ func (fkc *FKCascadeExec) buildFKCascadePlan(ctx context.Context) (plannercore.P
...
-	if !ok {
-		return nil, nil
-	}
-	stmtNode, err := exec.ParseWithParams(ctx, sqlStr)
-	if err != nil {
-		return nil, err
-	}
 	ret := &plannercore.PreprocessorReturn{}
-	err = plannercore.Preprocess(ctx, sctx, stmtNode, plannercore.WithPreprocessorReturn(ret), plannercore.InitTxnContextProvider)
+	err := plannercore.Preprocess(ctx, sctx, stmtNode, plannercore.WithPreprocessorReturn(ret), plannercore.InitTxnContextProvider)
```
Comment (ekexium) [inline executor/foreign_key.go:680]: What's the point of `plannercore.InitTxnContextProvider` here? Seems to me the only usage of it is handling state read which we don't want here?

Comment (crazycs520) [reply executor/foreign_key.go:680]: @winoros PTAL

Comment (crazycs520) [reply executor/foreign_key.go:680]: Look like we can remove `plannercore.WithPreprocessorReturn(ret), plannercore.InitTxnContextProvider`, @winoros

File: executor/adapter.go [hunk 2]
```diff
@@ -611,6 +611,16 @@ func (a *ExecStmt) handleForeignKeyTrigger(ctx context.Context, e Executor, dept
 	return nil
 }
 
+// handleForeignKeyCascade uses to execute foreign key cascade behaviour, the progress is:
+//  1. Build delete/update executor for foreign key on delete/update behaviour.
+//     a. Construct delete/update AST. We used to try generated SQL string first and then parse the SQL to get AST,
+//     but we need convert Datum to string, there may be some risks here, so we chose to construct AST directly.
```
Comment (AilinKid) [inline executor/adapter.go:617]: any risk explanation comment here？

File: executor/foreign_key.go [hunk 3]
```diff
@@ -727,120 +718,100 @@ func (fkc *FKCascadeExec) fetchUpdatedValuesCouple() *UpdatedValuesCouple {
...
+
+func genWhereConditionAst(cols []*model.ColumnInfo, fkValues [][]types.Datum) ast.ExprNode {
+	if len(cols) > 1 {
+		return genWhereConditionAstForMultiColumn(cols, fkValues)
+	}
+	valueList := make([]ast.ExprNode, 0, len(fkValues))
+	for _, fkVals := range fkValues {
+		v := &driver.ValueExpr{Datum: fkVals[0]}
+		v.Type = cols[0].FieldType
+		row := &ast.ParenthesesExpr{Expr: v}
```
Comment (AilinKid) [inline executor/foreign_key.go:784]: can we eliminate the `ParenthesesExpr` wrapper?

File: executor/adapter.go (hunk 2 above)
Comment (you06) [reply executor/adapter.go:617]: The parse and restore function may not be a bijection.

```
Datum a = ...;
assert_eq(parse(restore(a))); -- this assertion may be broken
```

**Review by you06 — APPROVED** [review]

File: executor/foreign_key.go (hunk 3 above)
Comment (crazycs520) [reply executor/foreign_key.go:784]: Nice catch,  done.

File: executor/adapter.go (hunk 2 above)
Comment (crazycs520) [reply executor/adapter.go]: great, already add this into comment.

**Review by AilinKid — APPROVED** [review]

**Review by ekexium — APPROVED** [review]

**Pull request merged**